engine/access/rpc/backend/transaction_subscription_metadata.go

```diff
@@ -0,0 +1,315 @@
+//
+// If the reference block ID is unset, it attempts to retrieve it from storage.
+//
+// Parameters:
+//   - txReferenceBlockID: The reference block ID of the transaction.
+//
+// No errors expected during normal operations.
+func (tm *transactionSubscriptionMetadata) initTransactionReferenceBlockID(txReferenceBlockID flow.Identifier) error {
+	// Get referenceBlockID if it is not set
+	if txReferenceBlockID == flow.ZeroID {
+		tx, err := tm.backendTransactions.transactions.ByID(tm.txResult.TransactionID)
+		if err != nil {
+			return err
+		}
+		txReferenceBlockID = tx.ReferenceBlockID
+	}
+
+	tm.txReferenceBlockID = txReferenceBlockID
+
+	return nil
+}
+
+// initBlockInfo determines the block that contains the transaction and updates metadata accordingly.
+//
+// This function searches the transaction’s collection and its corresponding block, updating
+// relevant fields in the metadata object.
+//
+// Expected errors during normal operation:
+//   - `storage.ErrNotFound` if the collection or block cannot be retrieved.
+func (tm *transactionSubscriptionMetadata) initBlockInfo() error {
+	collection, err := tm.collections.LightByTransactionID(tm.txResult.TransactionID)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			return nil
+		}
+
+		return err
+	}
+
+	tm.txResult.CollectionID = collection.ID()
+
+	block, err := tm.blocks.ByCollectionID(tm.txResult.CollectionID)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			return nil
+		}
+
+		return err
+	}
+
+	tm.blockWithTx = block.Header
+	tm.txResult.BlockID = block.ID()
+	tm.txResult.BlockHeight = block.Header.Height
+
+	return nil
+}
+
+// initTransactionResult initializes the transaction result.
+//
+// Parameters:
+//   - ctx: Context for managing the operation lifecycle.
+//
+// No errors expected during normal operations.
+func (tm *transactionSubscriptionMetadata) initTransactionResult(ctx context.Context) error {
+	if err := tm.refreshTransactionResult(ctx); err != nil {
+		return err
+	}
+
+	// It is possible to receive a new transaction status while searching for the transaction result or do not find the
+	// transaction result at all, that is why transaction status must be filled after searching the transaction result
+	if err := tm.refreshStatus(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Refresh updates the transaction subscription metadata to reflect the latest state.
+//
+// Parameters:
+//   - ctx: Context for managing the operation lifecycle.
+//   - height: The block height used for searching transaction data.
+//
+// Expected errors during normal operation:
+//   - `ErrBlockNotReady` if the block at the given height is not found.
+func (tm *transactionSubscriptionMetadata) Refresh(ctx context.Context, height uint64) error {
+	if err := tm.refreshCollection(height); err != nil {
+		return err
+	}
+
+	if err := tm.refreshBlock(); err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
```

Comment:
this check is hidden in `refreshCollection` but included here for blocks. let's be consistent and put the check in either the refresh method or here.